Sources/Parser.php

```diff
@@ -0,0 +1,762 @@
+		}
+
+		return $code;
+	}
+
+	/*************************
+	 * Internal static methods
+	 *************************/
+
+	/**
+	 * Sets the values of this class's static variables.
+	 *
+	 * If a variable already has a value, the existing value is not changed.
+	 * This ensures that custom values set by external code are respected.
+	 */
+	protected static function setStaticVars(): void
+	{
+		// Is anything disabled?
+		self::$enable_bbc = self::$enable_bbc ?? !empty(Config::$modSettings['enableBBC']);
+		self::$enable_post_html = self::$enable_post_html ?? !empty(Config::$modSettings['enablePostHTML']);
+		self::$enable_markdown = self::$enable_markdown ?? !empty(Config::$modSettings['enableMarkdown']);
+		self::$custom_smileys_enabled = self::$custom_smileys_enabled ?? !empty(Config::$modSettings['smiley_enable']);
+
+		// Set up localization.
+		if (!isset(User::$me)) {
+			User::setMe(0);
+		}
+
+		self::$time_offset = self::$time_offset ?? User::$me->time_offset ?? 0;
+		self::$time_format = self::$time_format ?? User::$me->time_format ?? Time::getTimeFormat();
+
+		self::$locale = self::$locale ?? Lang::$txt['lang_locale'] ?? '';
+		self::$encoding = self::$encoding ?? (!empty(Utils::$context['utf8']) ? 'UTF-8' : (!empty(Config::$modSettings['global_character_set']) ? Config::$modSettings['global_character_set'] : (!empty(Lang::$txt['lang_character_set']) ? Lang::$txt['lang_character_set'] : 'UTF-8')));
+
+		// Smiley settings.
+		self::$custom_smileys_enabled = self::$custom_smileys_enabled ?? !empty(Config::$modSettings['smiley_enable']);
+		self::$smileys_url = self::$smileys_url ?? Config::$modSettings['smileys_url'];
+		self::$smiley_set = self::$smiley_set ?? (!empty(User::$me->smiley_set) ? User::$me->smiley_set : (!empty(Config::$modSettings['smiley_sets_default']) ? Config::$modSettings['smiley_sets_default'] : 'none'));
+	}
+
+	/**
+	 * Fills in any missing elements of $options with the default values.
+	 *
+	 * @param array $options An array of parser options.
+	 * @return array An updated copy of $options.
+	 */
+	protected static function setOptions(array $options): array
+	{
+		IntegrationHook::call('integrate_parser_options', [&$options]);
+
+		return array_merge(self::$defalt_options, $options);
+	}
+
+	/**
+	 * Transforms the input string into HTML.
+	 *
+	 * @param string $string The string in which to transform markup.
+	 * @param int $input_types Bitmask of this class's INPUT_* constants.
+	 *    Only the indicated types of markup will be parsed in the input string.
+	 * @param array $options An array of parser options.
+	 * @return string The transformed string.
+	 */
+	protected static function toHTML(string $string, int $input_types, array $options): string
+	{
+		// Allow mods access before parsing.
+		$smileys = !empty($input_types & self::INPUT_SMILEYS);
+
+		IntegrationHook::call('integrate_pre_parsebbc', [&$string, &$smileys, &$options['cache_id'], &$options['parse_tags']]);
```

Comment:
Do we call the pre_parsebbc here or in the BBC logic?

Reply:
I moved these two hooks here so that hooked functions could have access to the input string before any changes were made and after all changes were made, which is what these hooks were expected to do back when BBCode parsing was the only type of parsing. If I had left them in `BBCodeParser::parse()`, hooked functions would have been called in the midst of the process, which would have had negative effects depending on what the hooked function tried to do.

If I were creating the hooks from scratch I would have named them `integrate_pre_parse` and `integrate_post_parse` (without the trailing `bbc`). But since the hooks already existed with these names, they continue to have those names.

Reply:
That makes sense in terms of keeping compatibility with older code.